nix: General improvements to the package

MOIS3Y opened the pull request:
Small changes in the method of fetch `gvc`.
This will simplify the update of the submodule if necessary,
I also changed the input arguments for `ignis.nix`. You no longer need to get them from `pkgs. ' every time.
Changes are also useful to find a more “clean” way to set::

- `PYTHONPATH`
- `GI_TYPELIB_PATH`
- `LD_LIBRARY_PATH`
- `GST_PLUGIN_PATH`
- `GDK_PIXBUF_MODULE_FILE`

I tried to use [buildPythonPackage](https://nixos.org/manual/nixpkgs/stable/#buildpythonpackage-function) with [wrapGApps*](https://nixos.org/manual/nixpkgs/stable/#ssec-gnome-plugins) but so far I have not found a solution.

Current changes do not affect the build phase, everything works in the same way.
I checked the work on simple bar from the documentation.

```console
./result/bin/ignis --version
Ignis 0.5.dev0
Branch: 
Commit: 375a7a3e7b8dfa877478d98516c4351481a3865d ()
```
```console
nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.12.21, NixOS, 25.05 (Warbler), 25.05.20250402.2c8d3f4`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.13`
 - nixpkgs: `/nix/store/fwhfa9pbx8vdi8nd5pcys665baz6xdxf-source`
````

MOIS3Y commented:
I have good news!

Finally, it is possible to build `ignis` using `buildPythonPackage` and remove`wrapProgram` to set `PYTHON_PATH`.

**Features:**
- `gvc` subproject comes from inputs.
This means to update the source is enough to run  `nix flake update gvc'.

- Used `flake-utils` and `systems` do not need to clutter up flake.nix with custom function `forAllSystems`
`github:nix-systems/default-linux` - exposes aarch64 and x86_64 for linux.
For more information:
[nix-systems](https://github.com/nix-systems/nix-systems)
[flake-utils](https://github.com/numtide/flake-utils)

Nix flake users can also quickly try `ignis`
```console
nix run "github:linkfrg/ignis/#ignis"
```

- You can add extra dependencies for the package using `extraPackages? [] `(not only Python libraries, executable files will also be available on runtime)

From `nix/ignis.nix`
```nix
  dependencies = extraPackages ++ [
    glib
    gtk4
    gtk4-layer-shell
    gobject-introspection
    dart-sass
    gstreamer
    libpulseaudio
    pipewire
    networkmanager
    gnome-bluetooth

    pygobject3
    pycairo
    click
    charset-normalizer
    markupsafe
    loguru
    certifi
    idna
    urllib3

  ];
```

I checked the work on the configuration [linkfrg dotfiles](https://github.com/linkfrg/dotfiles/tree/main/ignis)

`flake.nix`

```nix
# snip
    ignis = {
      url = "github:MOIS3Y/ignis?ref=nix";
      # ! impotant for override (NIX will not allow override dependencies if the input data does not follow your system pkgs)
      inputs.nixpkgs.follows = "nixpkgs";  
    };
# snip
```

`host-configuration.nix`
```nix
# snip
  environment.systemPackages = [
    (inputs.ignis.packages.${pkgs.stdenv.hostPlatform.system}.ignis.override {
      extraPackages = [
        pkgs.python312Packages.psutil
        pkgs.python312Packages.jinja2
        pkgs.python312Packages.pillow
        pkgs.python312Packages.materialyoucolor
      ];
 # snip
    })
  ];
```

```console
ignis --version
Ignis 0.5.dev0
Branch: 
Commit: b7f02b4bea53f9821191fbb885ccb7c825091466 ()
```
![изображение](https://github.com/user-attachments/assets/b4273ba8-c271-4b17-adc0-84a5a7b32bd7)

Unfortunately, I have not found a way to add automatically
GI_TYPELIB_PATH
LD_LIBRARY_PATH
GST_PLUGIN_PATH
GDK_PIXBUF_MODULE_FILE
They still need to be set as arguments for `makeWrapperArgs`

I found changes for the commit by studying the source code:
[mk-python-derivation.nix](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/interpreters/python/mk-python-derivation.nix)

@r03l sorry, I have not yet added `nixosModule' #250. It is better to add features gradually.

r03l commented:
Looks good! I can add the nixosmodule always later :)

linkfrg commented:
Awesome! I guess we have to mention these features in the documentation.
Let's create a separate page for Nix, e.g., `docs/user/nix.rst`.



> Unfortunately, I have not found a way to add automatically
> GI_TYPELIB_PATH
> LD_LIBRARY_PATH
> GST_PLUGIN_PATH
> GDK_PIXBUF_MODULE_FILE

Hmm, doesn't `wrapGAppsHook4` work? Here is the lutris package what uses it too (but for gtk3): [nixpkgs](https://github.com/NixOS/nixpkgs/blob/nixos-24.11/pkgs/applications/misc/lutris/default.nix)

MOIS3Y commented:
Yes `buildPhase` uses only meson
It sets thith place:
`format = "other";`

From src:

```nix
  # pyproject = true <-> format = "pyproject"
  # pyproject = false <-> format = "other"
  # https://github.com/NixOS/nixpkgs/issues/253154
  pyproject ? null,

  # Several package formats are supported.
  # "setuptools" : Install a common setuptools/distutils based package. This builds a wheel.
  # "wheel" : Install from a pre-compiled wheel.
  # "pyproject": Install a package using a ``pyproject.toml`` file (PEP517). This builds a wheel.
  # "egg": Install a package from an egg.
  # "other" : Provide your own buildPhase and installPhase.
  format ? null,
```
If set `format = "pyproject";` and `build-systems = [ meson ];`  it couldn't find `pyproject.toml`


I've watched lutris nix and think it might work

```nix
  # avoid double wrapping
  dontWrapGApps = true;
  makeWrapperArgs = [
    "--prefix PATH : ${lib.makeBinPath requiredTools}"
    "\${gappsWrapperArgs[@]}"
  ];

```
I'm not sure btw:). The difference is what they make the `buildPythonApplication` and we make `buildPythonPackage`.

From [nixpkgs manual](https://nixos.org/manual/nixpkgs/stable/#buildpythonapplication-function)
>The [buildPythonApplication](https://nixos.org/manual/nixpkgs/stable/#buildpythonapplication-function) function is practically the same as [buildPythonPackage](https://nixos.org/manual/nixpkgs/stable/#buildpythonpackage-function). The main purpose of this function is to build a Python package where one is interested only in the executables, and not importable modules. For that reason, when adding this package to a [python.buildEnv](https://nixos.org/manual/nixpkgs/stable/#python.buildenv-function), the modules won’t be made available.

We have both.
 We need to import modules in`~/.config/ignis`. In fact, we just expand in runtime the work of ignis exec file.
 Maybe we should  use `buildPythonApplication` ?
 
On nix, when you edit `~/.config/config.py` LSP server, doesn't see the imported modules (like #110) because when we add it to `/nix/store` `ignis` is not available for a system interpreter.
This is not a package problem, it's just how `nix` works. Everything is isolated.

I really like this idea #137 
If i`gnis` was used as a framework on the basis of which the user could create his own shells.
I mean so that the user can create his executable file by adding `ignis` as dependency.
`Astal` does so, providing libraries to write your desktop-shell.
The current approach with the executable file is certainly easier and clearer for a wide range of users.

MOIS3Y commented:
Yes, `wrapGAppsHook4` works with both `buildPythonPackage`, `buildPythonApplication`

<details>
  <summary>It looks cleaner</summary>
  
```nix
{ self
, lib
, wrapGAppsHook4
, pkg-config
, meson
, ninja
, git
, glib
, gtk4
, gtk4-layer-shell
, gobject-introspection
, librsvg
, dart-sass
, libpulseaudio
, pipewire
, networkmanager
, gnome-bluetooth
, python312Packages
, gst_all_1
, gvc
, extraPackages ? []
, version ? "git"
}:
  let
    inherit (lib)
      licenses
      platforms
      maintainers
    ;
    inherit (python312Packages)
      buildPythonApplication
      pygobject3
      pycairo
      click
      charset-normalizer
      markupsafe
      loguru
      certifi
      idna
      urllib3
    ;
    inherit (gst_all_1)
      gstreamer
      gst-plugins-base
      gst-plugins-good
      gst-plugins-bad
      gst-plugins-ugly
    ;
  in buildPythonApplication {

  inherit version;
  pname = "ignis";
  src = "${self}";

  format = "other";

  nativeBuildInputs = [
    pkg-config
    meson
    ninja
    git
    gobject-introspection
    wrapGAppsHook4
  ];

  dependencies = extraPackages ++ [
    glib
    gtk4
    gtk4-layer-shell
    gobject-introspection
    dart-sass
    gstreamer
    gst-plugins-base
    gst-plugins-good
    gst-plugins-bad
    gst-plugins-ugly
    librsvg
    libpulseaudio
    pipewire
    networkmanager
    gnome-bluetooth

    pygobject3
    pycairo
    click
    charset-normalizer
    markupsafe
    loguru
    certifi
    idna
    urllib3
  ];

  patchPhase = ''
    mkdir -p ./subprojects/gvc
    cp -r ${gvc}/* ./subprojects/gvc
  '';

  buildPhase = ''
    cd ..
    meson setup build --prefix=$out -DCOMMITHASH=${self.rev or "dirty"}
    ninja -C build
  '';

  installPhase = ''
    ninja -C build install
  '';

  #? avoid double wrapping. we manually pass args to wrapper
  dontWrapGApps = true;
  preFixup = ''
    makeWrapperArgs+=(
      "''${gappsWrapperArgs[@]}"
      --set LD_LIBRARY_PATH "$out/lib:${gtk4-layer-shell}/lib:$LD_LIBRARY_PATH"
    )
  '';

  meta = {
    description = ''
      A widget framework for building desktop shells,
      written and configurable in Python.
    '';
    homepage = "https://github.com/linkfrg/ignis";
    changelog = "https://github.com/linkfrg/ignis/releases";
    license = licenses.gpl3;
    platforms = platforms.linux;
    maintainers = with maintainers; [ frdiener somokill MOIS3Y ];
    mainProgram = "ignis";
  };
}
```
</details>

I also checked the work on [linkfrg dotfiles](https://github.com/linkfrg/dotfiles/tree/main/ignis)
Should I send this option?

linkfrg commented:
> Should I send this option?

You mean, commit this change? If so, yes, it looks good

MOIS3Y commented:
Done!

I've used `buildPythonPackage`.
This allows to add `ignis` to system Python as site-packages. It will close #110
I will explain below how it works.

How to use this flake?

Add `Ignis` to your flake's inputs:

```nix
ignis = {
  url = "github:MOIS3Y/ignis?ref=nix";
  # ! impotant for override (NIX will not allow override dependencies if the input data does not follow your system pkgs)
  inputs.nixpkgs.follows = "nixpkgs";  
};
```

Then add the following to `environment.systemPackages` or `home.packages`:

```nix
inputs.ignis.packages.${system}.ignis
```

You can even add this package to the system Python (here's an example for home-manager packages)

```nix
{ config, pkgs, inputs, ... }: {
  home.packages = with pkgs; [
    (python3.withPackages(ps: with ps; [
      inputs.ignis.packages.${pkgs.stdenv.hostPlatform.system}.ignis
      pip
      # add more python pkgs here ...
    ]))
    firefox
    neovim
    # add more hm pkgs here ..
  ];
}
```
This method will allow the LSP server of your favorite text editor to see the `ignis` modules.

 If you need to add additional dependencies, use the func `.overrire`.

`configuration.nix`

```nix
{ config, pkgs, lib, inputs, ... }: {
# snip ...
  environment.systemPackages = [
    (inputs.ignis.packages.${pkgs.stdenv.hostPlatform.system}.ignis.override {
      extraPackages = [
        pkgs.python312Packages.psutil
        pkgs.python312Packages.jinja2
        pkgs.python312Packages.pillow
        pkgs.python312Packages.materialyoucolor
        # add more dependencies here ...
      ];
    })
    firefox
    neovim
    # add more pkgs here ...
  ];
# snip ...
}
```

Or even like this:
```nix
{ config, pkgs, inputs, ... }: {
  # snip ...
  home.packages = with pkgs; [
    (python3.withPackages(ps: with ps; [
      (inputs.ignis.packages.${pkgs.stdenv.hostPlatform.system}.ignis.override {
        extraPackages = [
          pkgs.python312Packages.psutil
          pkgs.python312Packages.jinja2
          pkgs.python312Packages.pillow
          pkgs.python312Packages.materialyoucolor
        ];
      })
    ]))
  ];
  # snip ...
}
```

MOIS3Y commented:
>Let's create a separate page for Nix, e.g., docs/user/nix.rst.

The documentation is also ready.
Probably it is better to send it with a separate PR

linkfrg commented:
> The documentation is also ready.
> Probably it is better to send it with a separate PR

Since this PR adds a new feature, it's better to update the documentation here right away

MOIS3Y commented on docs/user/nix.rst:
```diff
@@ -0,0 +1,204 @@
+Nix
+============
+
+Introduction
+------------
+The manual describes the installation of Ignis on NixOS,
+but it is also applicable for installation on any other Linux distro
+with ``Nix`` package manager.
+You only need to adapt the ``flake.nix``.
+
+Supported systems: ``["x86_64-linux" "aarch64-linux"]``
```
I agree.
I wrote an introduction only for users who are slightly familiar with Nix.
It seemed to me that the manual below might seem applicable only for NixOS users.

MOIS3Y replied:
I did not find recommendations in the documentation of how to send corrections.
Should I make new commits of corrections or force push of the current commit with corrections?

Found this [PR](https://github.com/linkfrg/ignis/pull/181) here were the new commits.

linkfrg replied:
Make new commits (there is the "commit suggestion" button for this)

MOIS3Y commented on docs/user/nix.rst:
```diff
@@ -0,0 +1,204 @@
+Installation
+------------
+
+To install it you must have flake enabled and your NixOS configuration must be managed with flakes.
+See  `Flakes <https://nixos.wiki/wiki/Flakes>`_ for instructions on how to install and enable them on NixOS.
```
Ok, I will delete it.

MOIS3Y commented on docs/user/nix.rst:
```diff
@@ -0,0 +1,204 @@
+Output data can be added in different ways.
+In the example below, the home manager is installed as a NixOS module
+
+Don't forget to pass ``inputs``
+to ``extraSpecialArgs`` for Home Manager
+or ``specialArgs`` for NixOS host configuration
+
+.. code-block:: nix
+
+    {
+      outputs = { self, nixpkgs, home-manager, ... }@inputs: let
+        system = "x86_64-linux";
+        lib = nixpkgs.lib;
+        extraSpecialArgs = { inherit system inputs; };  # <- passing inputs to the attribute set for home-manager
+        specialArgs = { inherit system inputs; };       # <- passing inputs to the attribute set for NixOS
+      in {
+        nixosConfigurations = {
+          dummy-hostname = lib.nixosSystem {
+            modules = [
+              inherit specialArgs;  # <- this will make inputs available anywhere in the NixOS configuration
+              ./path/to/configuration.nix
+              home-manager.nixosModules.home-manager {
+                home-manager = {
+                  inherit extraSpecialArgs;  # <- this will make inputs available anywhere in the HM configuration
+                  useGlobalPkgs = true;
+                  useUserPackages = true;
+                  users.yourUserName = import ./path/to/home.nix;
+                };
+              }
+            ];
+          };
+        };
+      };
+    }
```
This is just a dummy example from the home-manager documentation.

In the last manual, we do not show to the user how an attribute `Inputs` will be available.
I mean that by default this attribute is not available.

If we think that the user is well acquainted with the `Nix Flakes`, we can simply write:

Don't forget to pass ``inputs``
to ``extraSpecialArgs`` for Home Manager
or ``specialArgs`` for NixOS host configuration

linkfrg replied:
Got it, I think we can move it to ``.. tip::`` block

MOIS3Y commented on docs/user/nix.rst:
```diff
@@ -0,0 +1,204 @@
+    it won't find extra dependencies
+
+
+Tips and tricks
+---------------
+
+Some services such as ``upower``, ``hyprland``
+expect that they are already present in the system.
+You need to take care of their launch yourself.
+
+For example, to use ``upower`` add the following to your configuration
+
+**configuration.nix**
+
+.. code-block:: nix
+
+    services.upower.enable = true;
+
+
```
Typically, comprehensive programs such as desktop-shell have special NixOS modules.
These modules will take care of starting the necessary services.

I agree, it is better to remove it.

linkfrg commented on docs/user/nix.rst:
```diff
@@ -0,0 +1,204 @@
+    }
+
+
+.. hint::
+    You can even add Ignis with extra dependencies to your system ``Python``
+
+It can be useful if the LSP server of your favorite text editor can't find Ignis modules
```
hm, interesting... so could we name this like `Adding Ignis to system Python`?

r03l replied:
Or adding to ignis to  python devshell / virtual environment? Because nixos advices you to not install python system wide https://nixos.wiki/wiki/Python

MOIS3Y replied:
`Adding Ignis to system Python` Yes, a good title.

We can move it to .. tip:: block too
And indicate that the recommended method is still ``environment.systemPackages`` or ``home.packages``

linkfrg commented:
I made some documentation improvements in 0a1a79c43a4a177759b6e1b0591c60854acb5445, you can check it out

linkfrg commented on nix/ignis.nix:
```diff
@@ -55,43 +104,27 @@ pkgs.stdenv.mkDerivation {
     license = licenses.gpl3;
-    maintainers = with maintainers; [ frdiener somokill ];
+    platforms = platforms.linux;
+    maintainers = with maintainers; [ frdiener somokill MOIS3Y ];
```
Are you sure you're going to actually maintain this package? By maintaining I mean regularly updating the package, fixing issues, etc.
I was planning to remove `frdiener` and `somokill` from the maintainers list, since they don't maintain this package in fact

MOIS3Y replied:
I can’t answer yet, I'm not sure that I have enough time.
If necessary, I will try to make edits if necessary.
Then let's just delete the attribute `maintainers`

linkfrg commented:
Looks good now... Is there anything to add, @MOIS3Y? Or it's ready to be merged?

linkfrg commented:
> Later, we can reconsider the structure of `/nix`
> It is easier to use just `default.nix` file at the root near `meson.build`
> This will allow us to change or not use `buildPhase` `installPhase` at all.
> Nix will do everything for us. We can simply use the `mesonFlags`.

I think we can do this in this PR (`shell.nix` should be added in a separate PR)

MOIS3Y commented:
- Removed unused attributes. From a Nix perspective, it's considered bad practice to keep unused attributes in expressions. The nixd linter flags them as errors.  
- Removed explicit `buildPhase` and `installPhase` - passing flags via `mesonFlags` is sufficient. The expression is now even cleaner.  
- Renamed `ignis.nix` to `default.nix` to follow Nix conventions. For now, I chose not to move `default.nix` to the project root. If we later add `shell.nix`, `nixosModule`, or `homeManagerModule`, it will be cleaner to keep them all in the same directory.

linkfrg commented:
@MOIS3Y, is it ready to be merged?

MOIS3Y commented:
Yes, the build is ready.

I'd also add a check attribute to flake.nix. This is useful for workflow automation. Running checks ensures the build passes for all declared architectures.

This would also allow setting up a GitHub Action in the future to automatically update `flake.lock`, keeping the main branch in sync with the latest `nixos-unstable` updates. If the build breaks, we'll notice immediately. And needless to say, if anything changes in the package's build algorithm itself, we'll be able to react quickly.

Here's how it might look:
```nix
        apps = rec {
          ignis = flake-utils.lib.mkApp { drv = self.packages.${system}.ignis;}
          #? workaround add meta attrs to avoid warning message
          // { meta = self.packages.${system}.ignis.meta; };
          default = ignis;
        };
        checks = self.packages.${system};
```
I'm unsure whether to keep the comment and workaround. The workaround doesn't affect the build—it just silences the warning.

Builds can be verified across all systems with:
```console
nix flake check --all-systems
```

Apologies for not suggesting all changes at once. I keep forgetting some obvious but convenient optimizations 😔

linkfrg reviewed (approved):
Looks good for me, awesome work! Thanks!
